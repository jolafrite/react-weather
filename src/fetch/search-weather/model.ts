export interface ISearchWeatherFetchResponse {
  message:	string;
  cod:	number
  count:	number
  list: ISearchWeather[]
}

export interface ISearchWeather {
  id: number;
  name: string;
  coord: {
    lat: number;
    lon: number;
  };
  main: {
    temp: number;
    pressure: number;
    humidity: number;
    temp_min: number;
    temp_max: number;
    sea_level: number;
    grnd_level: number;
    temp_kf: number;
  };
  dt: number;
  wind: {
    speed: number;
    deg: number;
  };
  sys: {
    country: string
  };
  rain: {
    "1h": number;
    "3h": number;
  };
  snow: {
    "1h": number;
    "3h": number;
  };
  clouds: {
    all: number;
  };
  weather: {
    id: number;
    main: string;
    description: string;
    icon: string
  }[];
}
