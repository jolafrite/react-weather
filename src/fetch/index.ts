export * from './load-weather/load-weather.fetch';
export * from './search-weather/search-weather.fetch';
