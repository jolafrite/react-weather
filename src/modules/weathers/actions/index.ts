import * as WeatherSearchPageActions from './weather-search-page.actions';
import * as WeatherViewPageActions from './weather-view-page.actions';
import * as WeatherApiActions from './weathers-api.actions';

export {
  WeatherSearchPageActions,
  WeatherViewPageActions,
  WeatherApiActions,
};
