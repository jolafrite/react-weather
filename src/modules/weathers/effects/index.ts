export * from './load-weather/load-weather.effect';
export * from './search-weather/search-weather.effect';
