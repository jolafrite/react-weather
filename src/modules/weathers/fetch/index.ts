export * from './load-5-days-weather-forecast/load-5-days-weather-forecast.fetch';
export * from './load-weather/load-weather.fetch';
export * from './search-weather/search-weather.fetch';
