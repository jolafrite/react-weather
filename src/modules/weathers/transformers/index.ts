export * from './search-weather/search-weather.transformer';
