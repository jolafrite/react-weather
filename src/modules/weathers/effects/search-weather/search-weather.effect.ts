import { WeatherSearchPageActions, WeatherApiActions } from '../../actions';
import { Dispatch } from 'redux';
import { searchWeatherFetch } from '../../../../fetch';

export const searchPageSearchWeatherEffect = (query: string) => async (dispatch: Dispatch) => {
  dispatch(WeatherSearchPageActions.search(query));

  return await searchWeathersEffect(
    query,
    WeatherSearchPageActions.searchSuccess,
    WeatherSearchPageActions.searchError,
    dispatch
  );
}

const searchWeathersEffect = async (query: string, onSuccess: any, onError: any, dispatch: Dispatch) => {
  try {
    const response = await searchWeatherFetch(query);

    dispatch(WeatherApiActions.loadListSuccess(response));
    dispatch(onSuccess(response));
  } catch(e) {
    dispatch(onError(e));
  }
};
