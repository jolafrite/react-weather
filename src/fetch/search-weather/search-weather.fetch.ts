import Axios from "axios";
import { IWeather } from "../../modules/weathers/models";
import { ISearchWeatherFetchResponse } from "./model";
import { searchWeatherListTransformer } from "../../modules/weathers/transformers";

export const searchWeatherFetch = async (query: string): Promise<IWeather[]> => {
  //qery must be minimum 3 characters long
  if(query.length < 3) {
    return [];
  };

  const { data } = await Axios.request<ISearchWeatherFetchResponse>({
    method: 'get',
    url: `${process.env.WEATHER_URL}/find`,
    params: {
      q: query,
      units: 'metric',
      appid: '{WEATHER_APPID}',
    }
  });

  return searchWeatherListTransformer(data.list);
};
