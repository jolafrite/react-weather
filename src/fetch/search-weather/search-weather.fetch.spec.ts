import { searchWeatherFetch } from "./search-weather.fetch";
import Axios from "axios";
import { searchWeatherListTransformer } from "../../modules/weathers/transformers";
import { cleanup } from "@testing-library/react";

jest.mock("axios");
jest.mock("../../modules/weathers/transformers");

describe('fetch: searchWeatherFetch', () => {

  afterEach(cleanup)

  describe('call with a query less that 3 characters', () => {
    test('should return an empty array', async () => {
      const result = await searchWeatherFetch('pa');

      expect(
        result
      ).toMatchSnapshot();
    });
  });

  describe('call with a query', () => {
    test('should call the api and return the transformed response', async () => {
      const axiosRequest = (Axios.request as jest.Mock<any>);

      axiosRequest.mockResolvedValueOnce({
        data: {
          list: 'the response'
        }
      });
      (searchWeatherListTransformer as jest.Mock<any>).mockReturnValue('the response transformed')

      const response = await searchWeatherFetch('paris');

      expect(
        axiosRequest.mock.calls
      ).toMatchSnapshot();

      expect(
        response
      ).toMatchSnapshot()
    });
  });
});
