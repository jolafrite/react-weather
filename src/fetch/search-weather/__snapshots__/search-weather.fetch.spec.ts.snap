// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch: searchWeatherFetch call with a query less that 3 characters should return an empty array 1`] = `Array []`;

exports[`fetch: searchWeatherFetch call with a query should call the api and return the transformed response 1`] = `
Array [
  Array [
    Object {
      "method": "get",
      "params": Object {
        "appid": "{WEATHER_APPID}",
        "q": "paris",
        "units": "metric",
      },
      "url": "undefined/find",
    },
  ],
]
`;

exports[`fetch: searchWeatherFetch call with a query should call the api and return the transformed response 2`] = `"the response transformed"`;
